inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -4157,32 +4157,47 @@ private final class EventListenerListenersSupplier<T extends EventListener> impl
         public Iterable<T> get(BeanResolutionContext beanResolutionContext) {
             if (listeners == null) {
                 List<T> listeners = new ArrayList<>(listenersDefinitions.size());
+                List<BeanRegistration<T>> registrations = new ArrayList<>(listenersDefinitions.size());
                 for (BeanDefinition<T> listenersDefinition : listenersDefinitions) {
-                    T listener;
+                    BeanRegistration<T> registration;
                     if (beanResolutionContext == null) {
                         try (BeanResolutionContext context = newResolutionContext(listenersDefinition, null)) {
                             try (BeanResolutionContext.Path ignored = context.getPath().pushEventListenerResolve(
                                 listenersDefinition,
                                 eventType
                             )) {
-                                listener = resolveBeanRegistration(context, listenersDefinition).bean;
+                                registration = resolveBeanRegistration(context, listenersDefinition);
                             }
                         }
                     } else {
                         try (BeanResolutionContext.Path ignored = beanResolutionContext.getPath().pushEventListenerResolve(
                             listenersDefinition,
                             eventType
                         )) {
-                            listener = resolveBeanRegistration(beanResolutionContext, listenersDefinition).bean;
+                            registration = resolveBeanRegistration(beanResolutionContext, listenersDefinition);
                         }
                     }
-                    listeners.add(listener);
+                    registrations.add(registration);
+                }
+                // Avoiding the simpler Comparator.comparing(BeanRegistration::getOrder) to keep lambda setup machinery
+                // off the critical startup path.
+                registrations.sort(BeanRegistrationOrderComparator.INSTANCE);
+                for (BeanRegistration<T> registration : registrations) {
+                    listeners.add(registration.getBean());
                 }
-                OrderUtil.sort(listeners);
```

Comment:
unclear to me why this fix is needed. `BeanRegistration` implements `Ordered`  which intializaes the order either from the definition or the bean itself. Why is this not working?

Reply:
Previously OrderUtil.sort() was being called on a list of the beans themselves, not the `BeanRegistration` objects. However it's a good point that a custom comparator isn't needed in this case, so I removed that.